extension/src/completion/autocompletionProvider.ts

```diff
@@ -171,30 +187,222 @@ function getMatchingWord(document: vscode.TextDocument, position: vscode.Positio
+    Local = vscode.CompletionItemKind.Variable,
+    Arg = vscode.CompletionItemKind.Value,
+    Defun = vscode.CompletionItemKind.Class,
+    Enum = vscode.CompletionItemKind.Constant,
+    Symbol = vscode.CompletionItemKind.Field,
+    Function = vscode.CompletionItemKind.Method
+}
+export namespace AutoLispKind {
+    export function vsKind(kind: AutoLispKind): vscode.CompletionItemKind { return kind as unknown as vscode.CompletionItemKind; }
+}
+
+export class CompletionSymbolCollector {
+    private values: Map<string, string[]> = new Map();
+    private used: string[] = [ ];
+    private ret: vscode.CompletionItem[] = [ ];
+    private word: string = "";
+    private rooted: boolean = false;
+    private previousItems: string[] = [ ];
+    constructor(){ // uses the Enum Names as the list values so we can F2 rename them without having to hunt down string versions of them
+        this.values.set(AutoLispKind[AutoLispKind.Arg], []);
+        this.values.set(AutoLispKind[AutoLispKind.Defun], []);
+        this.values.set(AutoLispKind[AutoLispKind.Global], []);
+        this.values.set(AutoLispKind[AutoLispKind.Local], []);
+        this.values.set(AutoLispKind[AutoLispKind.Symbol], []);
+    }
+
+    get allValues (): Map<string, string[]> { return this.values;}
+
+    private assembleResult(lst: string[], kind: vscode.CompletionItemKind){
+        const platform = (os.platform() === "win32") ? "WIN" : "MAC";
+        lst.forEach(name => {
+            const key = name.toLowerCase();
+            if (platform === "WIN" || !winOnlyListFuncPrefix.find(pref => key.startsWith(pref))){
+                if (this.used.indexOf(key) === -1 && (this.word === "" || key.startsWith(this.word))){
+                    this.ret.push(new vscode.CompletionItem(name, kind));
+                    this.used.push(key);
+                }
+            }
+        });
     }
-    let allSuggestions: Array<vscode.CompletionItem> = [];
-    allSuggestions = getCompletionCandidates(candidatesItems, word, isupper);
 
-    if (os.platform() === "win32") {
-        return allSuggestions;
+    private tryAddSymbol(name: string, kind: AutoLispKind): void {
+        const ucName = name.toUpperCase();
+        if (name.indexOf(' ') === -1 && !this.values.get(AutoLispKind[kind]).find(p => p.toUpperCase() === ucName)){
+            this.values.get(AutoLispKind[kind]).push(name);
+        }
+        this.previousItems.unshift(ucName);
     }
-    else {
-        return allSuggestions.filter(function(suggestion) {
-            for (var prefix of winOnlyListFuncPrefix) {
-                if (suggestion.label.startsWith(prefix)) {
-                    return false;
+
+    private traverseSexpression(sx: Sexpression): void {
+        let context: number|string = "";
+        sx.atoms.forEach(atom =>{
+            if (context !== "EXIT" && atom instanceof Sexpression){    
+                this.traverseSexpression(atom as Sexpression);
+                if (context === 2){
```

Comment:
I don't understand this variable, what's the mean for 2 or 1?

Reply:
That is fair and I considered making this an enum multiple times. If you don't understand then I certainly need to fix it. Essentially _Context_ is a dynamic state flag that talks in strings for DEFUN's and integers for SETQ's. If you think about a setq, it is a series of nested VarName (1) = Value (2).